### PR TITLE
When searching for an ingester to send chunks to, ensure we can actually connect to it.

### LINKED ISSUE #1184: Failed to transfer chunks to another ingester: rpc error: code = Unavailable desc = there is no connection available
```
level=info ts=2019-01-11T20:17:39.047328696Z caller=transfer.go:195 msg="sending chunks" to_ingester=10.52.9.93:9095
level=error ts=2019-01-11T20:17:39.047697551Z caller=lifecycler.go:450 msg="Failed to transfer chunks to another ingester" err="rpc error: code = Unavailable desc = there is no connection available"
```

### PR DESCRIPTION
**Pull request by tomwilkie**
Fixes #1184 

Signed-off-by: Tom Wilkie <tom.wilkie@gmail.com>

### CODE REVIEW

**Comment by tomwilkie** [comment]
Doesn't seem to fix it:
```
level=info ts=2019-01-12T09:47:01.33001897Z caller=transfer.go:197 msg="sending chunks" to_ingester=10.52.27.27:9095
INFO: 2019/01/12 09:47:01 parsed scheme: ""
INFO: 2019/01/12 09:47:01 scheme "" not registered, fallback to default scheme
INFO: 2019/01/12 09:47:01 ccResolverWrapper: sending new addresses to cc: [{10.52.27.27:9095 0  <nil>}]
INFO: 2019/01/12 09:47:01 balancerWrapper: got update addr from Notify: [{10.52.27.27:9095 <nil>}]
level=error ts=2019-01-12T09:47:01.330539903Z caller=lifecycler.go:450 msg="Failed to transfer chunks to another ingester" err="rpc error: code = Unavailable desc = there is no connection available"
```

**Comment by tomwilkie** [comment]
Scratch that - never actually checking the err.  Thats why late night friday coding is not a good idea.

**Comment by tomwilkie** [comment]
Yeah works nicely with this:
```
level=info ts=2019-01-12T10:47:36.740986791Z caller=gokit.go:36 msg="=== received SIGINT/SIGTERM ===\n*** exiting"
level=info ts=2019-01-12T10:47:36.741103067Z caller=lifecycler.go:441 msg="changing ingester state from" old_state=ACTIVE new_state=LEAVING
INFO: 2019/01/12 10:48:06 parsed scheme: ""
INFO: 2019/01/12 10:48:06 scheme "" not registered, fallback to default scheme
level=info ts=2019-01-12T10:48:06.767099053Z caller=transfer.go:198 msg="sending chunks" to_ingester=10.52.9.218:9095
INFO: 2019/01/12 10:48:06 parsed scheme: ""
INFO: 2019/01/12 10:48:06 scheme "" not registered, fallback to default scheme
INFO: 2019/01/12 10:48:06 ccResolverWrapper: sending new addresses to cc: [{10.52.9.218:9095 0  <nil>}]
INFO: 2019/01/12 10:48:06 ClientConn switching balancer to "pick_first"
INFO: 2019/01/12 10:48:06 pickfirstBalancer: HandleSubConnStateChange: 0xc4b05ba990, CONNECTING
INFO: 2019/01/12 10:48:06 pickfirstBalancer: HandleSubConnStateChange: 0xc4b05ba990, READY
level=info ts=2019-01-12T10:48:16.499741057Z caller=transfer.go:252 msg="successfully sent chunks" to_ingester=10.52.9.218:9095
level=info ts=2019-01-12T10:48:16.523740024Z caller=lifecycler.go:339 msg="ingester removed from consul"
level=info ts=2019-01-12T10:48:16.523836343Z caller=lifecycler.go:263 msg="Ingester.loop() exited gracefully"
```

**Comment by bboreham** [comment]
Seems like this could create a future TOCTTOU issue. Why not look for another ingester if we get a failure to transfer?

**Comment by tomwilkie** [comment]
> Seems like this could create a future TOCTTOU issue. Why not look for another ingester if we get a failure to transfer?

That issue already exists, right? This change slightly mitigates it, but I agree it is not perfect.  We currently fallback to flushing, and I'd prefer to implement the "proper" migration ideas than drastically change the current approach, which I think it a bit of a dead end.

**Comment by bboreham** [comment]
> That issue already exists, right?

Not the one I meant.  Currently we pick an ingester based on the list in Consul and try to transfer, and if that fails we're done.  There is no 'check' so no time-of-check issue.

You think it's way more complicated to put a loop round the pick+transfer set?

**Comment by tomwilkie** [comment]
Ah okay I see, sorry.  I'll see what a retry loop looks like.

**Comment by tomwilkie** [comment]
@bboreham PTAL

**Comment by gouthamve** [comment]
LGTM from me!

File: pkg/ring/lifecycler.go [hunk 1]
```diff
@@ -30,6 +30,11 @@ var (
 		Name: "cortex_ingester_ring_tokens_to_own",
 		Help: "The number of tokens to own in the ring.",
 	})
+	shutdownDuration = promauto.NewHistogramVec(prometheus.HistogramOpts{
+		Name:    "cortex_shutdown_duration_seconds",
+		Help:    "Duration (in seconds) of cortex shutdown procedure (ie transfer or flush).",
+		Buckets: prometheus.DefBuckets,
```
Comment (bboreham) [inline pkg/ring/lifecycler.go]: This tops out at 10 seconds?  My ingesters take much longer than that to transfer or flush.

File: pkg/ingester/transfer.go [hunk 2]
```diff
@@ -266,19 +287,30 @@ func (i *Ingester) findTargetIngester(ctx context.Context) (*ring.IngesterDesc,
...
+	defer ticker.Stop()
+
 	for {
-		ingester, err := findIngester()
-		if err != nil {
-			level.Debug(util.Logger).Log("msg", "Error looking for pending ingester", "err", err)
-			if time.Now().Before(deadline) {
-				time.Sleep(i.cfg.SearchPendingFor / pendingSearchIterations)
+		select {
+		case <-ticker.C:
```
Comment (bboreham) [inline pkg/ingester/transfer.go]: If we have an outer loop doing retries, do we still want this inner loop?

Comment (tomwilkie) [reply pkg/ingester/transfer.go]: I've also deprecated one of the flags that was used here (`ingester.search-pending-for`), which is replaced by `ingester.max-transfer-retries`.  I made the max backoff 5s, so with 10 retries this should approximate the default 10s search time from the previous flag.

**Comment by bboreham** [comment]
PLEASE PUT FLAG MEANING CHANGES LOUDLY IN THE TITLE

**Pull request merged**